TheThingsNetwork/lorawan-stack
# Rate limiting Redis backend

#### Summary



Ref https://github.com/TheThingsIndustries/lorawan-stack/issues/2969

#### Changes



- Add Redis store to rate limiting configuration
- Conditionally use Redis store for rate limiting

#### Testing



Tested locally by adding rate limiting rules and checking Redis keys.

##### Test steps:

- Add rate limiting to `config/stack/ttn-lw-stack.yml` configuration from [rate limiting docs](https://www.thethingsindustries.com/docs/reference/rate-limiting/)
  - Set the rate limits of "User login" (`http:account`) low so you can reproduce hitting the limit (`max-per-min: 3`)
  - Use `redis` rate limiting provider
    - ```yaml
      rate-limiting:
        provider: 'redis'
        profiles:
          - name: User login
            max-per-min: 3
            associations:
              - http:account
      ```
- Follow [Getting started steps from `DEVELOPMENT.md`](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.29/DEVELOPMENT.md#getting-started) to setup the stack locally
- Execute the start command
  - `$ go run ./cmd/ttn-lw-stack -c ./config/stack/ttn-lw-stack.yml start`
- Open `redis-cli` on the Redis container while the stack is running to observe Redis keys
  - `$ docker exec -it lorawan-stack-dev-redis-1 redis-cli`
- List Redis keys in `redis-cli`
  - `127.0.0.1:6379> KEYS ttn:v3:rate-limiting:*`
  - There should be **no keys** listed
- Open the Console in the browser **on the login page**
  - [http:localhost:1885/console](http:localhost:1885/console)
  - If you've previously been logged in you need to log out.
- Check the Redis keys again
  - `127.0.0.1:6379> KEYS ttn:v3:rate-limiting:*`
  - There should be **multiple keys** listed.
- Refresh the page multiple times to hit the rate limit.
  - When the limit is set to 3 you should be rate-limited on the 3rd attempt since you've loaded the page for the first time when opening it.
- Wait a minute since the last refresh and check the keys again
  - `127.0.0.1:6379> KEYS ttn:v3:rate-limiting:*`
  - There should be **no keys** listed

##### Regressions



Check if in-memory rate limiting is working correctly.

#### Notes for Reviewers



This PR was moved from enterprise https://github.com/TheThingsIndustries/lorawan-stack/pull/4071

#### Checklist
